PLAT-5306 Push artifacts to Jfrog

tapansstardog opened the pull request:
@pdmars I have added the files to integrate with Jenkins. I have tried to mimic the behavior of publishing of Stardog artifacts. There may be some code that is not needed for JedAI artifacts. Request you to have a look at the files and changes. 
I could not test them as I am not sure how to make it visible on Jenkins portal. Need your help in enabling it. This will help me in debugging my changes.

tapansstardog committed:
Initial commit for Jenkins integration

tapansstardog committed:
Missed Jenkinsfile file

tapansstardog committed:
removed commented code and fixed build issues

pdmars commented on .ops/jenkins/delete-nightly-snapshot.sh:
```diff
@@ -0,0 +1,7 @@
+             --user ${artifactoryUsername} \
+             --password ${artifactoryPassword}  \
+             --quiet=true \
+             --spec=./ops/jenkins/snapshot/delete.spec || true
```
If you haven't, you can try this locally and I strongly suggest you try it with the dry run option to see what it'll delete.

tapansstardog replied:
Yes, I haven't done it yet. Will do it tomorrow. Will my own credentials work in this case (write permission needed)? I tried to publish it using scala build script it returned with authentication errors.

pdmars replied:
I'm not sure but if you run into issues we can sort them out. Authentication errors to me sounds more like something wasn't setup correctly with the creds, not that they weren't authorized for certain permissions.

tapansstardog replied:
I will look into it again, Paul.

tapansstardog replied:
I had to make a little change in artifactory url in build.sbt. Both publishing to jfrog and deleting from it, are working fine.

pdmars commented on Jenkinsfile:
```diff
@@ -0,0 +1,121 @@
+pipeline {
+    agent {
+        dockerfile {
+            dir '.ops/build-docker'
```
This is referencing items in the main stardog repo, but this is it's own repo.

I don't believe the jenkinsfile can depend on items from another repo like this since the location of this file is tied with the repo.

In cases where the stardog pipeline uses other repos, all the base support is in the stardog repo and then we clone the needed repos and run from there. That's one option we can do here (e.g. keep this jenkinsfile and whatnot in the stardog repo like we do for starbench but then have this pipeline clone the jedai repo and run whatever tests/scripts you want).

It's also fine to have this as it's own thing in this repo, but then you will need to add your own stripped down dockerfile that builds everything you need for this stuff.

tapansstardog replied:
> This is referencing items in the main stardog repo, but this is it's own repo.

My bad. I should have seen that. Missed it as haven't run it through Jenkins yet.

This repo does not depend on Stardog so I believe we may not need all the stuff that we have under Dockerfile of Stardog repo. Since starbench depends on Stardog so it would make sense to have base stuff in Stardog itself. 

However, in our case (jedai-spark), I think we can create a trimmed down version of Dockerfile. Thoughts?

pdmars replied:
Yep, that all makes sense to me. A slimmed down version of what you need sounds good. I think going forward we'll be basing a lot of our CI images on the latest LTS ubuntu docker images -- official decision is still pending -- that are pretty slim (then you can install what you need). So basing it on something like the `ubuntu:22.04` docker image (assuming that works with what you need) will likely be in line with stuff we do in other places of CI going forward just FYI.

tapansstardog replied:
I have added the slimmed down Dockerfile (with ubuntu:22.04) and tested locally on my system. It is installing the necessary libraries. I could not test the code on Jenkins as I couldn't see "Build with parameters" option. I tried to make the build parameterized but did not see that on Jenkins either.

tapansstardog replied:
Looks like it failed to compile as `JAVA_VERSION` is the `choice` parameter and there is no choice provided while defining it. Making necessary changes.

tapansstardog replied:
@pdmars I have made the necessary changes. Jenkins build is running successfully and uploading artifacts on jfrog. I verified by adding its dependency in the separate project.

I have also moved `jenkins` folder from `.ops` to `ops`. I found this folder structure in `stardog` repository so did this change.

Couple of queries:
I think we do not need `stardog-testing/nightly-develop-jedai-snapshot/binaries/` directory when all the required artifacts are uploaded under `MAVEN` folder. I can remove the code associated with this.

Secondly, in `Dockerfile`, I found `openjdk` installed under `/usr/lib/jvm/java-11-openjdk-arm64` while in case of Jenkins it was `/usr/lib/jvm/java-11-openjdk-amd64`. Couldn't figure out.

pdmars replied:
Yeah no need to do anything with `nightly-develop-jedai-snapshot/binaries/` if we are not using it. Just uploading the maven artifacts, if that's all you need, is fine.

tapansstardog replied:
Removed that code.

pdmars reviewed (changes requested):
I added this pipeline to jenkins if you want to keep it in it's own repo like this:
https://jenkins.dev.stardog.com/job/stardog-jedai/

Or we can put the jenkins support for these items in the stardog repo, which is what I'd recommend if anything in this project will depend on anything in stardog more than just downloading a release to deploy/use. If not, having it separate like this is fine but you'll need to remove these items that reference stuff in the stardog repo and replace them with separate implementations here.

tapansstardog committed:
Added Dockerfile to run ubuntu container.
Tested delete-nightly-snapshot locally.
Tested build.sbt changes for uploading artifacts to jfrog.

tapansstardog committed:
made corrections and updated Dockerfile

tapansstardog committed:
Removed the reference to pre-build.sh in Jenkinsfile

tapansstardog committed:
Moved jenkins folder from .ops to ops.
Removed choice parameter of Java 11.

tapansstardog committed:
Updated jar file to uploaded under binaries.

pdmars commented on .ops/build-docker/Dockerfile:
```diff
@@ -0,0 +1,88 @@
+ARG GROUP_ID=1001
+
+ENV DEFAULT_JAVA_VER 11
+ENV JAVA_HOME /usr/lib/jvm/java-${DEFAULT_JAVA_VER}-openjdk-arm64/
```
this should be amd64 i think. jenkins isn't an arm system

tapansstardog replied:
Yeah, made it amd64

pdmars commented on .ops/build-docker/Dockerfile:
```diff
@@ -0,0 +1,88 @@
+
+RUN apt-get -qq update && \
+    apt-get install -y software-properties-common && \
+    add-apt-repository ppa:openjdk-r/ppa && \
```
does ubuntu 22.04 still need this repo to get java 11? that may just be the default ubuntu ships in its repos

tapansstardog replied:
removed the repo

pdmars commented on .ops/build-docker/Dockerfile:
```diff
@@ -0,0 +1,88 @@
+    sudo \
+    python2.7 \
+    python3-pip \
+    openjdk-8-jdk \
```
do you need java 8?

tapansstardog replied:
When I removed java 8 from here, I kept on getting the following error. I tried various changes while searching the solution on the net but couldn't succeed.  Can you suggest what's the missing dependency it is looking for? openjdk-8 must be installing some dependencies required for ca-certificates.

```
head: cannot open '/etc/ssl/certs/java/cacerts' for reading: No such file or directory
dpkg: error processing package ca-certificates-java (--configure):
 installed ca-certificates-java package post-installation script subprocess returned error exit status 1
dpkg: dependency problems prevent configuration of openjdk-11-jre-headless:amd64:
 openjdk-11-jre-headless:amd64 depends on ca-certificates-java (>= 20190405~); however:
  Package ca-certificates-java is not configured yet.
```

pdmars replied:
there's the `ca-certificates` being installed in this line below. Maybe try moving that one up to it's own line and installing it before this? java 8 maybe brought it in but java 11 isn't for some reason. I'm not sure.

pdmars commented on .ops/build-docker/Dockerfile:
```diff
@@ -0,0 +1,88 @@
+    apt-get -y install \
+    docker-compose \
+    sudo \
+    python2.7 \
```
do you need python 2.7?

tapansstardog replied:
Removed all references of python

pdmars commented on .ops/build-docker/Dockerfile:
```diff
@@ -0,0 +1,88 @@
+# RUN locale-gen en_US.UTF-8 && \
+#     echo 'LANG="en_US.UTF-8"' > /etc/default/locale
+
+RUN pip3 install awscli --upgrade
```
are you using the aws cli?

tapansstardog replied:
No, Paul. Removed this as well

pdmars commented on Jenkinsfile:
```diff
@@ -0,0 +1,105 @@
+        TF_VAR_snapshot_url = 'http://stardog.jfrog.io/stardog/stardog-snapshots-local'
+        TF_VAR_artifactory_username = credentials('artifactoryUsername')
+        TF_VAR_artifactory_password = credentials('artifactoryPassword')
+        STARDOG_GITHUB_API_TOKEN = credentials('stardog-github-api-token')
```
I'm guessing this is just copy/pasta, but it's not clear we need these. we can probably just add the ones you are using.

tapansstardog replied:
I have removed them.

pdmars commented on Jenkinsfile:
```diff
@@ -0,0 +1,105 @@
+    parameters {
+        booleanParam(name: 'CREATE_RELEASE_ZIP', defaultValue: false, description: 'Create the release zip from the dist directory')
+        string(name: 'AWS_INSTANCE_TYPE', defaultValue: 'm4.xlarge', description: 'AWS instance type for the databricks, chaos, or release tests (m4.4xlarge or larger is recommended if dataset is >1m)')
+        string(name: 'AWS_INSTANCE_CIDR', defaultValue: '127.0.0.1/32', description: 'The CIDR that will have access to Stardog running on 5820 in the environment for databricks, chaos, or release tests (if applicable)')
```
remove these options if we don't need them? fine if you do, just didn't see them used in the PR

tapansstardog replied:
I have removed most of them, however kept a couple of these parameters expecting to use them .

pdmars commented on .ops/build-docker/Dockerfile:
```diff
@@ -0,0 +1,88 @@
+RUN file="$(ls -1 /usr/lib/jvm/)" && echo $file
+RUN echo $(ls -1 /usr/lib/jvm/)
+
+RUN ln -sf /usr/lib/jvm/java-11-openjdk-arm64/bin/java /etc/alternatives/java
```
i think amd64 here too, on jenkins we do:

```
RUN update-java-alternatives --set java-1.11.0-openjdk-amd64
RUN ln -s ${JAVA_HOME} /usr/lib/jvm/default-java
```

tapansstardog replied:
Yes, made the changes

pdmars commented on Jenkinsfile:
```diff
@@ -0,0 +1,105 @@
+    echo "Checking Java version"
+    echo "JAVA_HOME=${env.JAVA_HOME}"
+    sh "java -version"
+}
```
I don't think we need all this java stuff, unless you need to add support for 8. But otherwise we can just use the default of 11.

tapansstardog replied:
Yes, I had it for debugging the installation. Now removed

pdmars reviewed (changes requested):
Just some cleanup suggestions with the setup

tapansstardog committed:
Incorporated PR suggestions

tapansstardog committed:
Removed more unwanted code from Jenkins

tapansstardog committed:
More cleanup

pdmars commented on .ops/build-docker/Dockerfile:
```diff
@@ -0,0 +1,81 @@
+ENV DEFAULT_JAVA_VER 11
+ENV JAVA_HOME /usr/lib/jvm/java-${DEFAULT_JAVA_VER}-openjdk-amd64/
+ENV DEBIAN_FRONTEND noninteractive
+# ENV LC_ALL en_US.UTF-8
```
remove?

tapansstardog replied:
Yes

pdmars commented on .ops/build-docker/Dockerfile:
```diff
@@ -0,0 +1,81 @@
+
+RUN apt-get -qq update && \
+    apt-get install -y software-properties-common && \
+    add-apt-repository ppa:ts.sch.gr/ppa && \
```
it might be worth figuring out what package comes from this repo and dropping it + the package if we don't need it.

tapansstardog replied:
Moved `ca-certificates` before java 11. `software-propoerties-common` is part of ubuntu Jammy image. Dropped other stuff.

pdmars commented on .ops/build-docker/Dockerfile:
```diff
@@ -0,0 +1,81 @@
+RUN useradd stardog --shell /bin/bash --create-home
+
+RUN file="$(ls -1 /usr/lib/jvm/)" && echo $file
+RUN echo $(ls -1 /usr/lib/jvm/)
```
is this debugging that we can remove when you're done with it?

tapansstardog replied:
Removed

pdmars commented on Jenkinsfile:
```diff
@@ -0,0 +1,84 @@
+
+        LANG = 'C.UTF-8'
+        S3_LOCAL_HOSTNAME = "s3${env.BUILD_ID}${UUID.randomUUID().toString().replace("-", "")}"
+        RHEL_CREDS = credentials('RHEL')
```
drop these, the s3 and rhel creds?

tapansstardog replied:
Done

pdmars commented on Jenkinsfile:
```diff
@@ -0,0 +1,84 @@
+def setJavaVersion() {
+    echo "Setting Java version to Java 11"
+    env.JAVA_HOME = "/usr/lib/jvm/java-11-openjdk-amd64/"
+    sh "sudo update-java-alternatives --set java-1.11.0-openjdk-amd64"
```
java 11 is the default in the image so you shouldn't need this either

tapansstardog replied:
Agreed

pdmars commented on Jenkinsfile:
```diff
@@ -0,0 +1,84 @@
+                    setJavaVersion()
+                }
+            }
+        }
```
drop this stage, I think

tapansstardog replied:
Done

tapansstardog committed:
PR suggestions incorporated

pdmars commented on .ops/build-docker/Dockerfile:
```diff
@@ -0,0 +1,70 @@
+    groff \
+    lcov \
+    bash-completion \
+    mingw-w64-x86-64-dev && \
```
there might be some others in here you can still pare back to keep this small: docker-compose, docker.io in particular, but otherwise this is looking much better (and smaller)!

tapansstardog replied:
Sure Paul. Will do that in next commit.

pdmars reviewed (approved):
Looks good to me. Your call on if there's anything else in the image you want to try to remove to keep it as small as possible, but I won't hold it up anymore over it.
